File: native_library/apple/FirebaseANE/FirestoreANE/FirestoreController.swift
/*
 *  Copyright 2018 Tua Rua Ltd.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import Foundation
import FreSwift
import Firebase
import FirebaseFirestore

class FirestoreController: FreSwiftController {
    static var TAG = "FirestoreController"
    internal var context: FreContextSwift!

    private var firestore: Firestore?
    private var batch: WriteBatch?
    private var snapshotRegistrations = [String: ListenerRegistration]()
    
    convenience init(context: FreContextSwift, loggingEnabled: Bool, settings: FirestoreSettings?) {
        self.init()
        self.context = context
        guard let app = FirebaseApp.app() else {
            warning(">>>>>>>>>> NO FirebaseApp !!!!!!!!!!!!!!!!!!!!!")
            return
        }
        if loggingEnabled {
            FirebaseConfiguration.shared.setLoggerLevel(.debug)
        }
        firestore = Firestore.firestore(app: app)
        if let settings = settings {
            firestore?.settings = settings
        }
    }
    
    func getFirestoreSettings() -> FREObject? {
        return firestore?.settings.toFREObject()
    }
    
    // MARK: - Documents
    
    func getDocuments(path: String, callbackId: String, whereList: [Where], orderList: [Order],
                      startAtList: [Any], startAfterList: [Any], endAtList: [Any],
                      endBeforeList: [Any], limitTo: Int) {
        guard let fs = firestore else {
            return
        }
        var q: Query = fs.collection(path)
        for w in whereList {
            switch w.operatr {
            case "==":
                q = q.whereField(w.fieldPath, isEqualTo: w.value!)
            case "<":
                q = q.whereField(w.fieldPath, isLessThan: w.value!)
            case ">":
                q = q.whereField(w.fieldPath, isGreaterThan: w.value!)
            case ">=":
                q = q.whereField(w.fieldPath, isGreaterThanOrEqualTo: w.value!)
            case "<=":
                q = q.whereField(w.fieldPath, isLessThanOrEqualTo: w.value!)
            default:
                break
            }
        }

        for o in orderList {
            q = q.order(by: o.by, descending: o.descending)
        }

        if !startAtList.isEmpty {
            q = q.start(at: startAtList)
        }
        if !startAfterList.isEmpty {
            q = q.start(after: startAfterList)
        }
        if !endAtList.isEmpty {
            q = q.end(at: endAtList)
        }
        if !endBeforeList.isEmpty {
            q = q.end(before: endBeforeList)
        }
        
        q = q.limit(to: limitTo)
        
        q.getDocuments { querySnapshot, error in
            if let err = error as NSError? {
                self.dispatchEvent(name: DocumentEvent.QUERY_SNAPSHOT,
                               value: DocumentEvent(
                                callbackId: callbackId,
                                error: err).toJSONString()
                )
                
            } else {
                if let qSnapshot: QuerySnapshot = querySnapshot {
                    self.dispatchEvent(name: DocumentEvent.QUERY_SNAPSHOT,
                                   value: DocumentEvent(
                                    callbackId: callbackId,
                                    data: qSnapshot.toDictionary()).toJSONString()
                    )
                }
 
            }
        }
        
    }
    
    func initDocumentReference(path: String) -> String? {
        return firestore?.document(path).documentID
    }
    
    func documentWithAutoId(path: String) -> String? {
        return firestore?.collection(path).document().path
    }
    
    func deleteDocumentReference(path: String, callbackId: String?) {
        let docRef = firestore?.document(path)
        docRef?.delete { error in
            if callbackId == nil { return }
            if let err = error as NSError? {
                self.dispatchEvent(name: DocumentEvent.DELETED,
                               value: DocumentEvent(callbackId: callbackId,
                                                    data: ["path": path],
                                                    error: err).toJSONString())
            } else {
                self.dispatchEvent(name: DocumentEvent.DELETED,
                               value: DocumentEvent(callbackId: callbackId,
                                                    data: ["path": path]).toJSONString())
            }
        }
    }
    
    func getDocumentParent(path: String) -> String? {
        return firestore?.document(path).parent.path
    }
    
    func addSnapshotListenerDocument(path: String, callbackId: String, callbackCallerId: String) {
        guard let docRef = firestore?.document(path) else { return }
        snapshotRegistrations[callbackCallerId] = docRef.addSnapshotListener({ snapshot, error in
            if let err = error as NSError? {
                self.dispatchEvent(name: DocumentEvent.SNAPSHOT,
                                   value: DocumentEvent(callbackId: callbackId,
                                                        data: nil,
                                                        realtime: true,
                                                        error: err).toJSONString()
                )
                
            } else {
                if let document = snapshot {
                    self.dispatchEvent(name: DocumentEvent.SNAPSHOT,
                                       value: DocumentEvent(callbackId: callbackId,
                                                             data: document.toDictionary(),
                                                             realtime: true).toJSONString()
                    )
                }
            }
        })
    }
    
    func removeSnapshotListener(callbackId: String) {
        snapshotRegistrations[callbackId]?.remove()
        snapshotRegistrations.removeValue(forKey: callbackId)
    }
    
    func getDocumentReference(path: String, callbackId: String) {
        let docRef = firestore?.document(path)
        docRef?.getDocument { snapshot, error in
            if let err = error as NSError? {
                self.dispatchEvent(name: DocumentEvent.SNAPSHOT,
                               value: DocumentEvent(callbackId: callbackId,
                                                    error: err).toJSONString())
            } else {
                if let document = snapshot {
                    self.dispatchEvent(name: DocumentEvent.SNAPSHOT,
                                   value: DocumentEvent(callbackId: callbackId,
                                                        data: document.toDictionary()).toJSONString()
                    )
                }
            }
        }   
    }
    
    func setDocumentReference(path: String, callbackId: String?, documentData: [String: Any], merge: Bool) {
        guard let docRef: DocumentReference = firestore?.document(path) else {
            return
        }
        docRef.setData(documentData, merge: merge, completion: { error in
            if callbackId == nil { return }
            if let err = error as NSError? {
                self.dispatchEvent(name: DocumentEvent.SET,
                               value: DocumentEvent(callbackId: callbackId,
                                                    data: ["path": path],
                                                    error: err).toJSONString())
            } else {
                self.dispatchEvent(name: DocumentEvent.SET,
                               value: DocumentEvent(callbackId: callbackId,
                                                    data: ["path": path]).toJSONString())
                
            }
        })
    }
    
    func updateDocumentReference(path: String, callbackId: String?, documentData: [String: Any]) {
        guard let docRef: DocumentReference = firestore?.document(path) else {
            return
        }
        docRef.updateData(documentData, completion: { error in
            if callbackId == nil { return }
            if let err = error as NSError? {
                self.dispatchEvent(name: DocumentEvent.UPDATED,
                               value: DocumentEvent(callbackId: callbackId,
                                                    data: ["path": path],
                                                    error: err).toJSONString())
            } else {
                self.dispatchEvent(name: DocumentEvent.UPDATED,
                               value: DocumentEvent(callbackId: callbackId,
                                                    data: ["path": path]).toJSONString())
            }
        })
    }
    
    // MARK: - Collections
 
    func getCollectionParent(path: String) -> String? {
        return firestore?.collection(path).parent?.path
    }
    
    func initCollectionReference(path: String) -> String? {
        return firestore?.collection(path).collectionID
    }
    
    // MARK: - Batch
    
    func startBatch() {
        batch = firestore?.batch()
    }
    
    func setBatch(path: String, documentData: [String: Any], merge: Bool) {
        if let docRef: DocumentReference = firestore?.document(path) {
            batch?.setData(documentData, forDocument: docRef, merge: merge)
        }
    }
    
    func updateBatch(path: String, documentData: [String: Any]) {
        if let docRef: DocumentReference = firestore?.document(path) {
           batch?.updateData(documentData, forDocument: docRef)
        }
    }
    
    func deleteBatch(path: String) {
        if let docRef: DocumentReference = firestore?.document(path) {
            batch?.deleteDocument(docRef)
        }
    }
    
    func commitBatch(callbackId: String?) {
        batch?.commit { error in
            if let err = error as NSError? {
                if callbackId == nil { return }
                self.dispatchEvent(name: BatchEvent.COMPLETE,
                               value: BatchEvent(callbackId: callbackId,
                                                 error: err).toJSONString())
                
            } else {
                self.dispatchEvent(name: BatchEvent.COMPLETE,
                               value: BatchEvent(callbackId: callbackId).toJSONString())
            }
        }
    }
    
    // MARK: - Network
    
    func enableNetwork(callbackId: String?) {
        firestore?.enableNetwork(completion: { error in
            if callbackId == nil { return }
            if let err = error as NSError? {
                self.dispatchEvent(name: NetworkEvent.ENABLED,
                               value: NetworkEvent(callbackId: callbackId, error: err).toJSONString())
            } else {
                self.dispatchEvent(name: NetworkEvent.ENABLED,
                               value: NetworkEvent(callbackId: callbackId).toJSONString())
            }
        })
    }
    
    func disableNetwork(callbackId: String?) {
        firestore?.disableNetwork(completion: { error in
            if callbackId == nil { return }
            if let err = error as NSError? {
                self.dispatchEvent(name: NetworkEvent.DISABLED,
                               value: NetworkEvent(callbackId: callbackId, error: err).toJSONString())
            } else {
                self.dispatchEvent(name: NetworkEvent.DISABLED,
                               value: NetworkEvent(callbackId: callbackId).toJSONString())
            }
        })
    }
    
}

File: native_library/android/FirebaseANE/Firestore/src/main/java/com/tuarua/firebase/firestore/FirestoreController.kt
/*
 *  Copyright 2018 Tua Rua Ltd.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package com.tuarua.firebase.firestore

import android.util.Log
import com.adobe.fre.FREContext
import com.google.firebase.firestore.*
import com.google.gson.Gson
import com.tuarua.firebase.firestore.data.Order
import com.tuarua.firebase.firestore.data.Where
import com.tuarua.frekotlin.FreException
import com.tuarua.frekotlin.FreKotlinController
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.Query.Direction.*
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.QuerySnapshot
import com.google.firebase.firestore.WriteBatch
import com.google.firebase.ktx.Firebase
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.app
import com.tuarua.firebase.firestore.events.*
import com.tuarua.firebase.firestore.extensions.*

class FirestoreController(override var context: FREContext?, loggingEnabled: Boolean,
                          settings: FirebaseFirestoreSettings?) : FreKotlinController {
    private lateinit var firestore: FirebaseFirestore
    private var snapshotRegistrations: MutableMap<String, ListenerRegistration> = HashMap()
    private val gson = Gson()
    private var batch: WriteBatch? = null

    init {
        try {
            FirebaseFirestore.setLoggingEnabled(loggingEnabled)
            firestore = Firebase.firestore(Firebase.app)
            if (settings != null) {
                firestore.firestoreSettings = settings
            }
        } catch (e: FreException) {
            warning(e.message)
            warning(e.stackTrace)
        } catch (e: Exception) {
            warning(e.message)
            warning(Log.getStackTraceString(e))
        }
    }

    fun getFirestoreSettings(): FirebaseFirestoreSettings = firestore.firestoreSettings

    /**************** Collections ****************/

    fun initCollectionReference(path: String): String = firestore.collection(path).id

    fun getCollectionParent(path: String): String? = firestore.collection(path).parent?.path

    /**************** Documents ****************/
    fun getDocuments(path: String, callbackId: String, whereList: MutableList<Where>,
                     orderList: MutableList<Order>, startAtList: List<Any>,
                     startAfterList: List<Any>, endAtList: List<Any>,
                     endBeforeList: List<Any>, limitTo: Int) {
        var q: Query = firestore.collection(path)
        for (w in whereList) when (w.operator) {
            "==" -> q = q.whereEqualTo(w.fieldPath, w.value)
            "<" -> q = q.whereLessThan(w.fieldPath, w.value)
            ">" -> q = q.whereGreaterThan(w.fieldPath, w.value)
            ">=" -> q = q.whereGreaterThanOrEqualTo(w.fieldPath, w.value)
            "<=" -> q = q.whereLessThanOrEqualTo(w.fieldPath, w.value)
        }

        orderList.forEach { o -> q = q.orderBy(o.by, if (o.descending) DESCENDING else ASCENDING) }

        if (startAtList.isNotEmpty()) {
            val startAtArr = startAtList.toTypedArray()
            q = q.startAt(*startAtArr)
        }
        if (startAfterList.isNotEmpty()) {
            val startAfterArr = startAfterList.toTypedArray()
            q = q.startAfter(*startAfterArr)
        }
        if (endAtList.isNotEmpty()) {
            val endAtArr = endAtList.toTypedArray()
            q = q.endAt(*endAtArr)
        }
        if (endBeforeList.isNotEmpty()) {
            val endBeforeArr = endBeforeList.toTypedArray()
            q = q.endBefore(*endBeforeArr)
        }
        q = q.limit(limitTo.toLong())

        q.get().addOnCompleteListener { task ->
            if (task.isSuccessful) {
                val qSnapshot: QuerySnapshot = task.result ?: return@addOnCompleteListener
                val docChanges = qSnapshot.documentChanges
                dispatchEvent(DocumentEvent.QUERY_SNAPSHOT, gson.toJson(DocumentEvent(callbackId, data = mapOf(
                        "metadata" to mapOf(
                                "isFromCache" to qSnapshot.metadata.isFromCache,
                                "hasPendingWrites" to qSnapshot.metadata.hasPendingWrites()),
                        "documentChanges" to docChanges.map { it.toMap() },
                        "documents" to qSnapshot.map { it.toMap() })))
                )
            } else {
                val error = task.exception as FirebaseFirestoreException
                dispatchEvent(DocumentEvent.QUERY_SNAPSHOT, gson.toJson(
                        DocumentEvent(callbackId, error = error.toMap()))
                )
            }
        }
    }

    fun initDocumentReference(path: String): String = firestore.document(path).id
    fun documentWithAutoId(path: String): String = firestore.collection(path).document().path

    fun getDocumentReference(path: String, callbackId: String) {
        val docRef: DocumentReference = firestore.document(path)
        docRef.get().addOnCompleteListener { task ->
            val documentSnapshot: DocumentSnapshot = task.result ?: return@addOnCompleteListener
            if (task.isSuccessful) {
                dispatchEvent(DocumentEvent.SNAPSHOT, gson.toJson(
                        DocumentEvent(callbackId, documentSnapshot.toMap()))
                )
            } else {
                val error = task.exception as FirebaseFirestoreException
                dispatchEvent(DocumentEvent.SNAPSHOT, gson.toJson(
                        DocumentEvent(callbackId, error = error.toMap()))
                )
            }
        }
    }

    fun addSnapshotListenerDocument(path: String, callbackId: String, callbackCallerId: String) {
        val docRef: DocumentReference = firestore.document(path)
        snapshotRegistrations[callbackCallerId] = docRef.addSnapshotListener { documentSnapshot, error ->
            when (error) {
                null -> if (documentSnapshot != null) {
                    dispatchEvent(DocumentEvent.SNAPSHOT, gson.toJson(
                            DocumentEvent(callbackId, documentSnapshot.toMap(), true))
                    )
                }
                else -> {
                    dispatchEvent(DocumentEvent.SNAPSHOT, gson.toJson(
                            DocumentEvent(callbackId, error = error.toMap()))
                    )
                }
            }
        }
    }

    fun removeSnapshotListener(id: String) {
        snapshotRegistrations[id]?.remove()
        snapshotRegistrations.remove(id)
    }

    fun setDocumentReference(path: String, callbackId: String?, documentData: Map<String, Any>, merge: Boolean) {
        val docRef: DocumentReference = firestore.document(path)
        if (merge) docRef.set(documentData, SetOptions.merge()).addOnCompleteListener { task ->
            if (callbackId == null) return@addOnCompleteListener
            if (task.isSuccessful) {
                dispatchEvent(DocumentEvent.SET, gson.toJson(DocumentEvent(callbackId, mapOf("path" to path))))
            } else {
                val error = task.exception as FirebaseFirestoreException
                dispatchEvent(DocumentEvent.SET, gson.toJson(
                        DocumentEvent(callbackId, error = error.toMap()))
                )
            }
        } else docRef.set(documentData).addOnCompleteListener { task ->
            if (callbackId == null) return@addOnCompleteListener
            if (task.isSuccessful) {
                dispatchEvent(DocumentEvent.SET, gson.toJson(DocumentEvent(callbackId, mapOf("path" to path))))
            } else {
                val error = task.exception as FirebaseFirestoreException
                dispatchEvent(DocumentEvent.SET, gson.toJson(
                        DocumentEvent(callbackId, error = error.toMap()))
                )
            }
        }
    }

    fun updateDocumentReference(path: String, callbackId: String?, documentData: Map<String, Any>) {
        val docRef: DocumentReference = firestore.document(path)
        docRef.update(documentData).addOnCompleteListener { task ->
            if (callbackId == null) return@addOnCompleteListener
            if (task.isSuccessful) {
                dispatchEvent(DocumentEvent.UPDATED, gson.toJson(
                        DocumentEvent(callbackId, mapOf("path" to path))))
            } else {
                val error = task.exception as FirebaseFirestoreException
                dispatchEvent(DocumentEvent.UPDATED, gson.toJson(
                        DocumentEvent(callbackId, mapOf("path" to path), error = error.toMap()))
                )
            }
        }
    }

    fun deleteDocumentReference(path: String, callbackId: String?) {
        val docRef: DocumentReference = firestore.document(path)
        docRef.delete().addOnCompleteListener { task ->
            if (callbackId == null) return@addOnCompleteListener
            if (task.isSuccessful) {
                dispatchEvent(DocumentEvent.DELETED, gson.toJson(DocumentEvent(callbackId)))
            } else {
                val error = task.exception as FirebaseFirestoreException
                dispatchEvent(DocumentEvent.DELETED, gson.toJson(
                        DocumentEvent(callbackId, mapOf("path" to path), error = error.toMap()))
                )
            }
        }
    }

    fun getDocumentParent(path: String): String = firestore.document(path).parent.path


    /**************** Batch ****************/
    fun startBatch() {
        batch = firestore.batch()
    }

    fun setBatch(path: String, documentData: Map<String, Any>, merge: Boolean) {
        val docRef: DocumentReference = firestore.document(path)
        when {
            merge -> batch?.set(docRef, documentData, SetOptions.merge())
            else -> batch?.set(docRef, documentData)
        }
    }

    fun updateBatch(path: String, documentData: Map<String, Any>) {
        val docRef: DocumentReference = firestore.document(path)
        batch?.update(docRef, documentData)
    }

    fun deleteBatch(path: String) {
        val docRef: DocumentReference = firestore.document(path)
        batch?.delete(docRef)
    }

    fun commitBatch(callbackId: String?) {
        batch?.commit()?.addOnCompleteListener { task ->
            if (callbackId == null) return@addOnCompleteListener
            if (task.isSuccessful) {
                dispatchEvent(BatchEvent.COMPLETE, gson.toJson(BatchEvent(callbackId)))
            } else {
                val error = task.exception as FirebaseFirestoreException
                dispatchEvent(BatchEvent.COMPLETE, gson.toJson(
                        NetworkEvent(callbackId, error.toMap()))
                )
            }
        }
    }

    fun enableNetwork(callbackId: String?) {
        firestore.enableNetwork().addOnCompleteListener { task ->
            if (callbackId == null) return@addOnCompleteListener
            if (task.isSuccessful) {
                dispatchEvent(NetworkEvent.ENABLED, gson.toJson(NetworkEvent(callbackId)))
            } else {
                val error = task.exception as FirebaseFirestoreException
                dispatchEvent(NetworkEvent.ENABLED, gson.toJson(
                        NetworkEvent(callbackId, error.toMap()))
                )
            }
        }
    }

    fun disableNetwork(callbackId: String?) {
        firestore.disableNetwork().addOnCompleteListener { task ->
            if (callbackId == null) return@addOnCompleteListener
            if (task.isSuccessful) {
                dispatchEvent(NetworkEvent.DISABLED, gson.toJson(NetworkEvent(callbackId)))
            } else {
                val error = task.exception as FirebaseFirestoreException
                dispatchEvent(NetworkEvent.DISABLED, gson.toJson(
                        NetworkEvent(callbackId, error.toMap()))
                )
            }
        }
    }

    override val TAG: String?
        get() = this::class.java.simpleName


}